src/persist-client/src/internal/encoding.rs

```diff
@@ -540,40 +538,109 @@ where
             trace: Some(self.collections.trace.into_proto()),
         }
     }
+}
+
+/// A decoded version of [ProtoStateRollup] for which we have not yet checked
+/// that codecs match the ones in durable state.
+#[derive(Debug)]
+pub struct UntypedState<T> {
```

Comment:
So, here's a couple things that I would be tempted to try out:
- Rename this to `Rollup` or similar. (The name feels appropriate since it's more or less 1:1 to the proto struct, and it saves us having so many structs named `*State`.)
- Replace the `into_proto` methods on `State` / `TypedState` with `into_rollup` methods, and have this guy be the only one that's converted back / forth to proto. Nice only because it makes the layering very clear. (Though I'm not sure whether/not that would require extra clones.)
- Move `shard_id` into this struct instead of `State`. `shard_id` is the odd one out in `State`, since it never changes. Moving it here would allow us to check that the decoded state is the state for the shard we expected at `check` time, and not carry around the `String`. (Though I guess some other folks will need a new `shard_id` field.)

I have not actually tried them, and this PR is certainly just fine without them. Odds are high that at least one of them is a bad idea. Feel free to explore or leave as you prefer.

Reply:
hm I do like the goal of further clarifying the name/intent, but I'm not sure about using `Rollup` in the name since it can also be forwarded to an arbitrary state via `apply_encoded_diffs` (not just a state used as a rollup).

I do like simplifying the proto serde layering, I found that a bit subtle while reading through

Reply:
okay, played around with all of these locally, but hit a snag with each of them. in general, I try pretty hard to say "yes" to all code review feedback, but there's enough small "gotchas" here (what you see is the result of quite a bit of time code golfing) that I might turn this around and counter with a request for a concrete diff

> Rename this to Rollup or similar. (The name feels appropriate since it's more or less 1:1 to the proto struct, and it saves us having so many structs named *State.)

I think there's a good case for naming it StateRollup to more strongly parallel the usual Foo <-> ProtoFoo naming. OTOH, I think that the Untyped* naming representing why _this_ variant is used (vs the other *State things) is probably the more important relationship to establish? if anything, maybe we'd want to rename ProtoStateRollup to ProtoUntypedState

> Replace the into_proto methods on State / TypedState with into_rollup methods, and have this guy be the only one that's converted back / forth to proto. Nice only because it makes the layering very clear. (Though I'm not sure whether/not that would require extra clones.)

yeah, the clones ended up killer here. I'm actually less stressed about the string clones in the (cold-ish) prod codepath (worst-case, we could make them `Cow<'a, str>`s) than I am about the State::clone we need for persistcli (I'd love to delete that method and couldn't find any way to do this without it)

> Move shard_id into this struct instead of State. shard_id is the odd one out in State, since it never changes. Moving it here would allow us to check that the decoded state is the state for the shard we expected at check time, and not carry around the String. (Though I guess some other folks will need a new shard_id field.)

agreed about shard_id being weird!

passing a shard_id to check ended up feeling a bit odd IMO because our only real option on a failure is to panic (either there or somewhere higher in the callstack in response to an error) and it wasn't clear to me that it was a win on the balance. also probably relevant is I kind of view these check things as a necessary evil (I _much_ perfer the code as it was before this PR) so adding more things to that set isn't something I'm particularly excited to do

Reply:
> in general, I try pretty hard to say "yes" to all code review feedback [...]

FWIW since my feedback was "please consider," and you did consider, I am satisfied! (And it's helpful to me to see your reasoning.)